src/contracts/core/DelegationManager.sol

```diff
@@ -224,48 +227,170 @@ contract DelegationManager is Initializable, OwnableUpgradeable, Pausable, Deleg
+        bool receiveAsTokens
+    ) external onlyWhenNotPaused(PAUSED_EXIT_WITHDRAWAL_QUEUE) {
+        bytes32 withdrawalRoot = calculateWithdrawalRoot(withdrawal);
+
+        require(
+            pendingWithdrawals[withdrawalRoot], 
+            "DelegationManager.completeQueuedAction: action is not in queue"
+        );
+
+        require(
+            slasher.canWithdraw(withdrawal.delegatedTo, withdrawal.startBlock, middlewareTimesIndex),
+            "DelegationManager.completeQueuedAction: pending action is still slashable"
+        );
+
+        require(
+            withdrawal.startBlock + withdrawalDelayBlocks <= block.number, 
+            "DelegationManager.completeQueuedAction: withdrawalDelayBlocks period has not yet passed"
+        );
+
+        require(
+            msg.sender == withdrawal.withdrawer, 
+            "DelegationManager.completeQueuedAction: only withdrawer can complete action"
+        );
+
+        if (receiveAsTokens) {
+            require(
+                tokens.length == withdrawal.strategies.length, 
+                "DelegationManager.completeQueuedAction: input length mismatch"
+            );
+        }
+
+        // Remove `withdrawalRoot` from pending roots
+        delete pendingWithdrawals[withdrawalRoot];
+
+        address currentOperator = delegatedTo[msg.sender];
+
+        // Finalize action by converting shares to tokens for each strategy, or
+        // by re-awarding shares in each strategy.
+        for (uint256 i = 0; i < withdrawal.strategies.length; ) {
+            if (receiveAsTokens) {
+                _withdrawSharesAsTokens({
+                    staker: withdrawal.staker,
+                    withdrawer: msg.sender,
+                    strategy: withdrawal.strategies[i],
+                    shares: withdrawal.shares[i],
+                    token: tokens[i]
+                });
+            } else {
+                // Award shares back to staker in StrategyManager/EigenPodManager
+                // If staker is delegated, increases shares delegated to operator
+                _addAndDelegateShares({
+                    staker: withdrawal.staker,
+                    withdrawer: msg.sender,
+                    operator: currentOperator,
+                    strategy: withdrawal.strategies[i],
+                    shares: withdrawal.shares[i]
+                });
+            }
+
+            unchecked { ++i; }
+        }
+
+        emit WithdrawalCompleted(withdrawalRoot);
+    }
+
+    // @notice Migrates an existing queued withdrawal from the StrategyManager contract to this contract.
+    function migrateQueuedWithdrawal(IStrategyManager.DeprecatedStruct_QueuedWithdrawal memory strategyManagerWithdrawalToMigrate) external {
+        // check for existence and delete the old storage
+        bytes32 oldWithdrawalRoot = strategyManager.calculateWithdrawalRoot(strategyManagerWithdrawalToMigrate);
+        strategyManager.migrateQueuedWithdrawal(oldWithdrawalRoot);
```

Comment:
might as well just pass the whole struct here if we're gonna calculate the withdrawal root by passing it the line above. the other option is to calculate in an internal function in the DM.

Reply:
> might as well just pass the whole struct here if we're gonna calculate the withdrawal root by passing it the line above.

That's exactly what I was doing before the request to only have the old root in the event.

>  the other option is to calculate in an internal function in the DM.

Hadn't considered this. Is having a public method on the StrategyManager still useful at all? Otherwise I do think this is the right approach.

Reply:
+1 for having the withdrawal be calculated in the DM for gas savings on migration. Doesn't appear that a public method is needed given other contracts don't have a dependency on this function

Reply:
Downside is additional bloat in the Dmgr contract.

Here's a solution that gives both, I think:
- Pass the struct to `Smgr.migrateQueuedWithdrawal`
- It returns the calculated root to the Dmgr

Reply:
Updated here: https://github.com/Layr-Labs/eigenlayer-contracts/pull/228

Reply:
LGTM. can close this thread out once the PR is merged.